liquidity-rewards/contracts/UniswapPoolRewards.sol

```diff
@@ -18,14 +18,14 @@ contract UniswapPoolRewards is PoolTokenWrapper, Ownable {
      * @dev We can set a reward rate per Ethereum block. 
      * Right now the average time per block is ~15 seconds, 
      * which is:
-     * 4 blocks per minute
+     * 4 blocks per minute = 0.0666... blocks/second
      * 240 blocks per hour (60 minutes)
      * 5,760 blocks per day (24 hours)
      * 172,800 blocks per month (30 days)
      * So we can use this to calculate a reward rate of 322,500 $ELIMU / 172,800 blocks/month = 1.87 $ELIMU/block 
-     * => rewardRate = 1.87 x 10e18 (token decimals) = 1870000000000000000
+     * => rewardRate = 1.87 ($ELIMU/block ) X 0.06666... (blocks/second) x 10e18 (token decimals) = 124666666666666666
      */
-    uint256 public rewardRate = 1870000000000000000;
+    uint256 public rewardRate = 124666666666666666;
```

Comment:
Does this mean that 0.125 `$ELIMU` is the reward per block?

Reply:
1.87 `$ELIMU` per block is equivalent to 0.125 `$ELIMU` per second according to our calculations.

Reply:
Isn't it easier to calculate the rewards based on the block heights instead of relying on time? I assumed each block will have a number value equal to the previous block +1? Let me know if I'm misunderstanding something fundamental here :-)

<img width="697" alt="Screen Shot 2022-07-14 at 4 25 31 PM" src="https://user-images.githubusercontent.com/15718174/178937477-bd7d4a5b-63a0-4e18-9735-6fbf8fd5ce3f.png">

Reply:
`block.number` value is presently a very unreliable method of measuring time (even for medium-term applications). Because, the time between blocks is already changing regularly due to the difficulty bomb.
Here we need the concept of time for updating the state variable `lastUpdateTime`, so we need a reward rate in the format of reward per second. Also, in the unit tests, it's more human-readable to test the contracts in time intervals.

Reply:
@hhio618 Thanks for explaining. What you are saying makes sense, I just haven't understood yet why it's necessary to do this using time (seconds) instead of block numbers. I'm looking at the [average block time chart](https://etherscan.io/chart/blocktime):

![Screen Shot 2022-07-15 at 11 17 53 AM](https://user-images.githubusercontent.com/15718174/179143280-04bcfade-f18e-4915-af2b-591be3cc6805.png)

It seems relatively stable between 13-15 seconds per block?

The reason I was suggesting to use block height instead of time as the unit of measurement is because I imagined that would make the code less complex. But I haven't actually tried coding it this way yet, so I don't know for sure how it would pan out.

Reply:
What are the reasons to prefer time-based calculations? I guess one reason to work with time-based calculations is to give a time frame to users for easily understanding the rewards allocation. For example, X tokens for an year or something. Any other reasons @nya-elimuai? However, as you have mentioned @hhio618, the issue is that block times are not very fixed. So if we go ahead with time calculations it is anyways going to drift a bit because the calculations are assuming a certain block time (here, `240 blocks per hour` is what we have assumed). 

Using block-based calculations implies that we can't always accurately claim that X tokens are allocated for an year because we are factoring in the fact that block times are going to be changing but that's an issue anyways.

So, I guess we should consider the ease of working with the contract. In which case would it be easier to work with (develop, test, debug, and maintain) - block-based or time-based? I, personally, think it would be easier is to probably test, debug and maintain because we have a block numbers to work with? We can switch from `lastRewardTime` to `lastRewardBlock`. 

What are your thoughts on this @hhio618?

Reply:
Thanks for sharing, @mitesh-mutha 🙂

This discussion made me realize that since the smart contract should be able to handle changes of the reward rate at any point in time, we should add unit tests using different values for the reward rate. So I added this issue for it: https://github.com/elimu-ai/web3-smart-contracts/projects/2#card-84206816

Reply:
@mitesh-mutha Thanks for your Infos. I must say that it's easier to work with the time-based approach, as it's hard to test the block-based approach as it seems that there is not a straightforward way to increase the `block.number` in the unit tests. also, in my opinion, a time-based approach is more realistic for testing different scenarios.